video/out/opengl/hwdec_ios.m

```diff
@@ -0,0 +1,295 @@
+        kCFAllocatorDefault,
+        NULL,
+        [EAGLContext currentContext],
+        NULL,
+        &p->gl_texture_cache);
+
+    if (err != noErr) {
+        MP_ERR(hw, "Failure in CVOpenGLESTextureCacheCreate: %d\n", err);
+        return -1;
+    }
+
+    p->vtctx = (struct mp_vt_ctx){
+        .priv = hw,
+        .get_vt_fmt = get_vt_fmt,
+    };
+    p->hwctx = (struct mp_hwdec_ctx){
+        .type = HWDEC_VIDEOTOOLBOX,
+        .download_image = download_image,
+        .ctx = &p->vtctx,
+    };
+    hwdec_devices_add(hw->devs, &p->hwctx);
+
+    return 0;
+}
+
+static int reinit(struct gl_hwdec *hw, struct mp_image_params *params)
+{
+    assert(params->imgfmt == hw->driver->imgfmt);
+
+    struct vt_format *f = vt_get_gl_format_from_imgfmt(params->hw_subfmt);
+    if (!f) {
+        MP_ERR(hw, "Unsupported CVPixelBuffer format.\n");
+        return -1;
+    }
+
+    params->imgfmt = f->imgfmt;
+    params->hw_subfmt = 0;
+    return 0;
+}
+
+static void cleanup_textures(struct gl_hwdec *hw)
+{
+    struct priv *p = hw->priv;
+    int i;
+
+    for (i = 0; i < MP_MAX_PLANES; i++) {
+        if (p->gl_planes[i]) {
+            CFRelease(p->gl_planes[i]);
+            p->gl_planes[i] = NULL;
+        }
+    }
+
+    CVOpenGLESTextureCacheFlush(p->gl_texture_cache, 0);
+}
+
+static int map_frame(struct gl_hwdec *hw, struct mp_image *hw_image,
+                     struct gl_hwdec_frame *out_frame)
+{
+    struct priv *p = hw->priv;
+    GL *gl = hw->gl;
+
+    CVPixelBufferRelease(p->pbuf);
+    p->pbuf = (CVPixelBufferRef)hw_image->planes[3];
+    CVPixelBufferRetain(p->pbuf);
+
+    uint32_t cvpixfmt = CVPixelBufferGetPixelFormatType(p->pbuf);
+    struct vt_format *f = vt_get_gl_format(cvpixfmt);
+    if (!f) {
+        MP_ERR(hw, "CVPixelBuffer has unsupported format type\n");
+        return -1;
+    }
+
+    const bool planar = CVPixelBufferIsPlanar(p->pbuf);
+    const int planes  = CVPixelBufferGetPlaneCount(p->pbuf);
+    assert(planar && planes == f->planes || f->planes == 1);
+
+    cleanup_textures(hw);
+
+    for (int i = 0; i < f->planes; i++) {
+        gl->ActiveTexture(GL_TEXTURE0 + i);
+
+        CVReturn err = CVOpenGLESTextureCacheCreateTextureFromImage(
+            kCFAllocatorDefault,
+            p->gl_texture_cache,
+            p->pbuf,
+            NULL,
+            GL_TEXTURE_2D,
+            f->gl[i].gl_internal_format,
+            CVPixelBufferGetWidthOfPlane(p->pbuf, i),
+            CVPixelBufferGetHeightOfPlane(p->pbuf, i),
+            f->gl[i].gl_format,
+            f->gl[i].gl_type,
+            i,
+            &p->gl_planes[i]);
+
+        if (err != noErr)
+            MP_ERR(hw, "error creating texture for plane %d: %d\n", i, err);
+
+        gl->BindTexture(CVOpenGLESTextureGetTarget(p->gl_planes[i]),
+                        CVOpenGLESTextureGetName(p->gl_planes[i]));
+        gl->TexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
```

Comment:
You use CVOpenGLESTextureGetTarget at first, but then you hardcode GL_TEXTURE_2D here. Also I'm not sure if the GL_TEXTURE_2D above fixes the target, or if the API is free to use another target (that would be most likely GL_TEXTURE_RECTANGLE - it would be nice if we could avoid it entirely, though the OSX equivalent still requires it AFAIK).